toolchain/check/merge.cpp

```diff
@@ -231,6 +231,27 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
     return false;
   }
 
+  if (new_param_pattern.Is<SemIR::AddrPattern>()) {
+    new_param_pattern = context.insts().Get(
+        new_param_pattern.As<SemIR::AddrPattern>().inner_id);
+    prev_param_pattern = context.insts().Get(
+        prev_param_pattern.As<SemIR::AddrPattern>().inner_id);
+    if (new_param_pattern.kind() != prev_param_pattern.kind()) {
+      emit_diagnostic();
+      return false;
+    }
+  }
+
+  auto new_name_id = SemIR::Function::GetNameFromPatternId(
+      context.sem_ir(), new_param_pattern_id);
+  auto prev_name_id = SemIR::Function::GetNameFromPatternId(
+      context.sem_ir(), prev_param_pattern_id);
```

Comment:
Small at this point, but could this go back to the prior:

```suggestion
  auto new_name_id = context.entity_names().Get(
      new_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id).name_id;
  auto prev_name_id = context.entity_names().Get(
      prev_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id).name_id;
```

i.e., because `GetNameFromPatternId` includes more logic, and if we don't need it, then this is efficient.